distribution/client/src/lib/update.ts

```diff
@@ -163,57 +198,76 @@ export default class Update {
        while it's not been yet marked as tainted in the backend?
      * how to report that borked case in a clear way
   */
-  restartJenkins(rollingBack?: boolean) { // Add param to stop recursion?
-    Supervisord.restartProcess('jenkins');
+  restartJenkins() {
 
-    const messageWhileRestarting = 'Jenkins should now be online, health checking!';
+    const jenkinsIsRestarting = Supervisord.restartProcess('jenkins');
+
+    const messageWhileRestarting = 'Jenkins is being restarted, health checking!';
     UI.publish(messageWhileRestarting);
     logger.info(messageWhileRestarting);
 
-    // FIXME: actually now I'm thinking throwing in HealthChecker might provide a more
-    // consistent promise usage UX here.
-    // checking healthState.health value is possibly a bit convoluted (?)
-    this.healthChecker.check()
-      .then( healthState => {
-        if (healthState.healthy) {
-          logger.info('Jenkins healthcheck after restart succeeded! Yey.');
-        } else {
-
-          // if things are wrong twice, stop trying and just holler for help
-          if (rollingBack) {
-
-            // Quick notice sketch, but I do think we need a very complete and informative message
-            const failedToRollbackMessage =
-              'Ooh noes :-(. We are terribly sorry but it looks like Jenkins failed to ' +
-              'upgrade, but even after the automated rollback we were unable to bring ' +
-              'to life. Please report this issue to the Jenkins Evergreen team. ' +
-              'Do not shutdown your instance as we have been notified of this failure ' +
-              'and are trying to understand what went wrong to push a new update that ' +
-              'will fix things.';
-            logger.error(failedToRollbackMessage);
-            UI.publish(failedToRollbackMessage);
-
-            // Not throwing an Error here as we want the client to keep running and ready
-            // since the next available UL _might_ fix the issue
-          } else {
-
-            const errorMessage = `Jenkins detected as unhealthy. Rolling back to previous update level (${healthState.message}).`;
-            UI.publish(errorMessage);
-            logger.warn(errorMessage);
-
-            this.snapshotter.revertToLevelBefore(this.getCurrentLevel());
-            this.revertToPreviousUpdateLevel();
-            this.restartJenkins(true);
-          }
-        }
+    return jenkinsIsRestarting
+      .then(() => this.healthChecker.check())
+      .then(() => {
+        logger.info('Jenkins healthcheck after restart succeeded! Yey.');
+        return true;
+
+      }).catch((error) => { // first catch, try rolling back
+
+        const errorMessage = `Jenkins detected as unhealthy. Rolling back to previous update level (${error}).`;
+        logger.warn(errorMessage);
+        UI.publish(errorMessage);
+
+        return this.revertToPreviousUpdateLevel(); // FIXME: async issue!
```

Comment:
This FIXME is a little scary to me, is this the follow up work for JEP-302, or is this a bug that still needs to be filed?